---
title: "Matilda_CMML_methylation_ChromHMM_2023"
author: "xiangfu"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(UpSetR)
library(GenomicRanges)

CMML_dataPath <- "/Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/"

DMP_data <- readr::read_delim(paste0(CMML_dataPath, "Matilda_data/DMPs_mC1_mC2_20230831.csv"), delim=";",locale=locale(decimal_mark = ",") )
meta_data <- readr::read_delim(paste0(CMML_dataPath, "Matilda_data/Meta_Data_20230831.csv"), delim=";" )

```

<h style="color:red">**UPDATE NOTE**</h>

- <h style="color:red">added additional dataset for DMPs with LFC>0.3  and LFC>0.2 according to the discussion on Tuesday 09.19</h>
- <h style="color:red">updated the enhancer gene association table, easy to use</h>



## filtering DMPs

filtering DMPs by `adj.P.Val < 0.01 & abs(logFC) >= 0.1`, 48148 DMPs left. Among them, 43104 sites were hypermethylated (logFC > 0); 5044 sits were hypomethylated (logFC < 0)

```{r}
## number of sites included in the received data
nrow(DMP_data)

## correct the comparison set
DMP_data$logFC <- DMP_data$logFC * -1

## subset significant DMPs
DMP_data.sig <- DMP_data %>% dplyr::filter(adj.P.Val < 0.01 & abs(logFC) >= 0.1) 

## number of significant DMPs
nrow(DMP_data.sig)

## hyper/hypo methylated sites
DMP_data.sig.hyper <- DMP_data.sig %>% dplyr::filter(logFC > 0) 
DMP_data.sig.hypo <- DMP_data.sig %>% dplyr::filter(logFC < 0) 

## double check 
DMP_data.sig %>% ggplot(aes(x=logFC, y=-log(adj.P.Val))) +geom_point()

# table(DMP_data.sig.hyper$CHR)
# table(DMP_data.sig.hypo$CHR)
# table(DMP_data.sig$CHR)
```

some numbers to help decide what logFC as cutoff

|logFC | DMP count | 
|:------:|:---------:|
|0.1 | 43104|
|0.15 | 22770|
|0.2 | 10960|
|0.25 | 4902|
|0.3 | 1943|

The table is based on filtering with (1) `adj.P.Val < 0.01` and (2) `!(mC_1_AVG < 0.1 & mC_2_AVG < 0.1)`, and then (3) `logFC`

```{r}
DMP_data %>% dplyr::mutate(sig=ifelse(.$adj.P.Val < 0.01, "sig","noSig")) %>% ggplot2::ggplot(aes(x=mC_1_AVG, y=mC_2_AVG, color=sig)) +geom_point(alpha=0.6,size=1)  # +geom_density2d()#

DMP_data %>% dplyr::filter(adj.P.Val < 0.01 )  %>% ggplot2::ggplot(aes(x=logFC)) +geom_density() +scale_x_continuous(breaks=seq(-0.4, 1, 0.1))

DMP_data %>% dplyr::filter(adj.P.Val < 0.01 ) %>% dplyr::filter(!(mC_1_AVG < 0.1 & mC_2_AVG < 0.1) )  %>% dplyr::filter(logFC >= 0.2 )  %>% nrow()

```

export significant sites to bed files

```{r}
DMP_data.sig %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end) %>% readr::write_delim(file=paste0(CMML_dataPath, "methyChromHMM/sigDMPs.bed"), delim="\t", col_names = FALSE)

DMP_data.sig.hyper %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end) %>% readr::write_delim(file=paste0(CMML_dataPath, "methyChromHMM/sigHyperDMPs.bed"), delim="\t", col_names = FALSE)

DMP_data.sig.hypo %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end) %>% readr::write_delim(file=paste0(CMML_dataPath, "methyChromHMM/sigHypoDMPs.bed"), delim="\t", col_names = FALSE)

```

## DMPs in Monocyte chromatin states

### chromatin state data retrieve 

we are using chromatin states from here [ChromHMM 18-state model for Monocytes-CD14+ RO01746 Primary Cells](https://www.encodeproject.org/annotations/ENCSR907LCD/). `ENCFF072SCA.bigBed` was downloaded. And convert it to `ENCFF072SCA.bed` by using [**bigBedToBed**](https://genome.ucsc.edu/goldenPath/help/bigBed.html). 

the model has 18 chromatin states as following:

<h style="color:red">filter out chrX/Y/M from RoadMap chromatin states</h>

```{r}
ENCFF072SCA.bed <- rtracklayer::import(paste0(CMML_dataPath, "RoadMap/ENCFF072SCA.bed"))
unique(ENCFF072SCA.bed$name)
length(ENCFF072SCA.bed)
ENCFF072SCA.bed <-  ENCFF072SCA.bed[seqnames(ENCFF072SCA.bed) %in% paste0("chr", unique(DMP_data$CHR))]
length(ENCFF072SCA.bed)

stateFactor <- lapply(unique(ENCFF072SCA.bed$name), function(x) str_split(x, "_")[[1]][1]) %>% unlist()  %>% as.numeric()
names(stateFactor) <- unique(ENCFF072SCA.bed$name)

ENCFF072SCA.bed %>% as.data.frame() %>%  dplyr::select(seqnames, name) %>% table() %>% as.data.frame() %>% dplyr::filter(!seqnames %in% c("chrX", "chrM")) %>% ggplot2::ggplot(aes(x=seqnames, y= name, fill=scale(Freq))) +geom_tile() + scale_fill_gradient(low = "white", high = "red") +theme(axis.text.x = element_text(angle = 45, hjust=1)) +xlab("") +ylab("") 
```

| STATE NO.  | MNEMONIC  |  DESCRIPTION |   |
|:---:|:---|:---:|:---:|
| 1  | TssA  |  Active TSS |   |
| 2  | TssFlnk  | Flanking TSS  |   |
| 3  | TssFlnkU  | Flanking TSS Upstream  |   |
| 4  | TssFlnkD  | Flanking TSS Downstream  |   |
| 5  | Tx  |  Strong transcription |   |
| 6  | TxWk  | Weak transcription  |   |
| 7  | EnhG1  | Genic enhancer1  |   |
| 8  | EnhG2  | Genic enhancer2  |   |
| 9  | EnhA1  |  Active Enhancer 1 |   |
| 10 | EnhA2 |  Active Enhancer 2 |   |
| 11  | EnhWk  | Weak Enhancer  |   |
| 12  | ZNF/Rpts  | ZNF genes & repeats  |   |
| 13  | Het  | Heterochromatin  |   |
| 14  | TssBiv |  Bivalent/Poised TSS |   |
| 15  | EnhBiv  | Bivalent Enhancer  |   |
| 16  | ReprPC  | Repressed PolyComb  |   |
| 17  | ReprPCWk  | Weak Repressed PolyComb  |   |
| 18  |  Quies | Quiescent/Low  |   |

[Summarized visualization of chromatin states](https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/core_K27ac/jointModel/final/annotationEnrichment_RoadmapEp_core_K27ac_18State.png)
	

### run chromHMM

```{bash, eval=FALSE}
cd /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM

### all significant DMPs
java -jar /Users/xiangfu.zhong/Documents/RESEARCH.KI/chromatinState/ChromHMM/ChromHMM.jar  OverlapEnrichment -labels /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/RoadMap/ENCFF072SCA.bed /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM/sigDMPs.bed  /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM/sigDMPs_within_chromatinStates

### hypermethylated DMPs
java -jar /Users/xiangfu.zhong/Documents/RESEARCH.KI/chromatinState/ChromHMM/ChromHMM.jar  OverlapEnrichment -labels /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/RoadMap/ENCFF072SCA.bed /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM/sigHyperDMPs.bed  /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM/sigHyperDMPs_within_chromatinStates

### hypomethylated DMPs
java -jar /Users/xiangfu.zhong/Documents/RESEARCH.KI/chromatinState/ChromHMM/ChromHMM.jar  OverlapEnrichment -labels /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/RoadMap/ENCFF072SCA.bed /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM/sigHypoDMPs.bed  /Users/xiangfu.zhong/Library/CloudStorage/OneDrive-KarolinskaInstitutet/Matilda_CMML_methylation_ChromHMM_2023/methyChromHMM/sigHypoDMPs_within_chromatinStates
```

### visualization

```{r message=FALSE, warning=FALSE}
sigDMPs.chromhmm <- readr::read_tsv( paste0( CMML_dataPath, "methyChromHMM/sigDMPs_within_chromatinStates.txt"), locale=locale(decimal_mark = ","))
colnames(sigDMPs.chromhmm) <- c("states", "Genome", "sigDMPs")

sigHyperDMPs.chromhmm <- readr::read_tsv( paste0( CMML_dataPath, "methyChromHMM/sigHyperDMPs_within_chromatinStates.txt"), locale=locale(decimal_mark = ","))
colnames(sigHyperDMPs.chromhmm) <- c("states"
                                     , "Genome", "sigHyperDMPs")

sigHypoDMPs.chromhmm <- readr::read_tsv( paste0( CMML_dataPath, "methyChromHMM/sigHypoDMPs_within_chromatinStates.txt"), locale=locale(decimal_mark = ","))
colnames(sigHypoDMPs.chromhmm) <- c("states", "Genome", "sigHypoDMPs")


pooledChromhmm  <- merge(sigDMPs.chromhmm[, c("states", "sigDMPs")], sigHyperDMPs.chromhmm[, c("states", "sigHyperDMPs")], by="states")
pooledChromhmm <- merge(pooledChromhmm, sigHypoDMPs.chromhmm[, c("states", "sigHypoDMPs")], by="states")
rownames(pooledChromhmm) <- pooledChromhmm$states
pooledChromhmm <- pooledChromhmm[,-1]

ComplexHeatmap::Heatmap(scale(pooledChromhmm), heatmap_legend_param = list(title = "Enrichment", title_position = "leftcenter-rot"), cluster_rows = FALSE, cluster_columns = FALSE,  col=circlize::colorRamp2(c(0,  1), c("white", "firebrick")), show_column_dend = FALSE, show_row_names = TRUE, row_names_side = "right", row_title = " ") 

```


### site proportion

[Tulstrup *et al.* 2021, Nature Communications](https://doi.org/10.1038/s41467-021-26093-2) they actually used site proportion, not a function from ChromHMM.

```{r}

#### function to get proportion of sites within chromatin states ####
getOverlapproportion <- function(queryDMPtab, setLabel, totalDMPtab, chrStateBed){
  states <- unique(chrStateBed$name)
  Nquery <- nrow(queryDMPtab)
  nonQueryTab <- totalDMPtab %>% dplyr::filter(!.$...1 %in% queryDMPtab$...1)
  NnonQuery <- nrow(nonQueryTab)
  
  queryBed <- queryDMPtab %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()
  nonQueryBed <- nonQueryTab %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()
  
  chromhmmProportion <- data.frame()
  
  for(istate in states){
    istateBed <- chrStateBed[chrStateBed$name==istate, ]
    # istateBed <- chrStateBed %>% as.data.frame() %>% dplyr::filter(name == istate)  %>% GenomicRanges::GRanges()
    
    queryCount <- GenomicRanges::findOverlaps(queryBed, istateBed) %>% as.data.frame() %>% dplyr::select(queryHits) %>% unlist() %>% unique() %>% length()
    nonQueryCount <- GenomicRanges::findOverlaps(nonQueryBed, istateBed) %>% as.data.frame() %>% dplyr::select(queryHits) %>% unlist() %>% unique() %>% length()
    
    chromhmmProportion <- rbind(data.frame(state=istate, label=setLabel, proportion=queryCount/as.double(Nquery), totalCount=as.double(Nquery)), chromhmmProportion)
    chromhmmProportion <- rbind(data.frame(state=istate, label=paste0("non",setLabel), proportion=nonQueryCount/as.double(NnonQuery), totalCount=as.double(NnonQuery)), chromhmmProportion)
  }
  return(chromhmmProportion)
}

#### significant DMPs ####
proportion.DMPsig <- getOverlapproportion(DMP_data.sig, "SigMethyled", DMP_data, ENCFF072SCA.bed)
proportion.DMPsig$state <- factor(proportion.DMPsig$state, levels = names(sort(-stateFactor))) 

proportion.DMPsig %>% ggplot2::ggplot(aes(x=proportion, y=state, fill=label)) +geom_bar(stat="identity", position="dodge") +xlab("proportion of sites") +ylab("Monocyte chromatin states") +ggtitle("significant DMPs") +scale_fill_manual(values=c("SigMethyled"="deeppink", "nonSigMethyled"="darkorange"))

##### significant hypermethyled sites ####
proportion.DMPsig.hyper <- getOverlapproportion(DMP_data.sig.hyper, "HyperMethyled", DMP_data, ENCFF072SCA.bed)
proportion.DMPsig.hyper$state <- factor(proportion.DMPsig.hyper$state, levels = names(sort(-stateFactor))) 

proportion.DMPsig.hyper %>% ggplot2::ggplot(aes(x=proportion, y=state, fill=label)) +geom_bar(stat="identity", position="dodge") +xlab("proportion of sites") +ylab("Monocyte chromatin states") +ggtitle("significant hypermethyled sites") +scale_fill_manual(values=c("HyperMethyled"="deeppink", "nonHyperMethyled"="darkorange"))

##### significant hypomethyled sites ####
proportion.DMPsig.hypo <- getOverlapproportion(DMP_data.sig.hypo, "HypoMethyled", DMP_data, ENCFF072SCA.bed)
proportion.DMPsig.hypo$state <- factor(proportion.DMPsig.hypo$state, levels = names(sort(-stateFactor))) 

proportion.DMPsig.hypo %>% ggplot2::ggplot(aes(x=proportion, y=state, fill=label)) +geom_bar(stat="identity", position="dodge") +xlab("proportion of sites") +ylab("Monocyte chromatin states") +ggtitle("significant hypomethyled sites") +scale_fill_manual(values=c("HypoMethyled"="deeppink", "nonHypoMethyled"="darkorange"))
```

####site proportion of merged enhancer set 

HyperMethyled site proportion of merged enhancer set:

enhancers:

- 7_EnhG1: Genic enhancer1     
- 8_EnhG2: Genic enhancer2
- 9_EnhA1: Active Enhancer 1
- 10_EnhA2: Active Enhancer 2
- 11_EnhWk: Weak Enhancer 

```{r}
enhancerBed <-  ENCFF072SCA.bed  %>% as.data.frame() %>% dplyr::filter(name %in% c("7_EnhG1", "8_EnhG2", "9_EnhA1", "10_EnhA2", "11_EnhWk" ))  %>% GenomicRanges::GRanges()

HyperMethyled.bed <- DMP_data.sig.hyper %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()

nonHyperMethyled.bed <- DMP_data %>% dplyr::filter(!.$...1 %in% HyperMethyled.bed$...1) %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()

hyperDMP_in_enhancer.count <- GenomicRanges::findOverlaps(HyperMethyled.bed, enhancerBed) %>% as.data.frame() %>% dplyr::select(queryHits) %>% unlist() %>% unique() %>% length()

nonHyperDMP_in_enhancer.count <- GenomicRanges::findOverlaps(nonHyperMethyled.bed, enhancerBed) %>% as.data.frame() %>% dplyr::select(queryHits) %>% unlist() %>% unique() %>% length()

# hyperDMP_in_enhancer.count/as.double(length(HyperMethyled.bed))
# nonHyperDMP_in_enhancer.count/as.double(length(nonHyperMethyled.bed))

sprintf("The proportion of hyper methyled sites (%s) within enhancers is: %s", length(HyperMethyled.bed), hyperDMP_in_enhancer.count/as.double(length(HyperMethyled.bed)))

sprintf("The proportion of nonhyper methyled sites (%s) within enhancers is: %s", length(nonHyperMethyled.bed), nonHyperDMP_in_enhancer.count/as.double(length(nonHyperMethyled.bed)))

```

#### enhancers have multiple hyper methyled sites

OUTPUT FILE: `multiple_HyperMethyled_in_enhancers.tsv`

```{r message=FALSE, warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

enhancer_buff <- data.frame()
for(iEnhType in c("7_EnhG1", "8_EnhG2", "9_EnhA1", "10_EnhA2", "11_EnhWk" )){
  iEnhType_enhancerBed <- enhancerBed[enhancerBed$name == iEnhType,]
  i_hyperDMP_in_enhancer <- GenomicRanges::findOverlaps(HyperMethyled.bed, iEnhType_enhancerBed)
  i_summaryTab <- subjectHits(i_hyperDMP_in_enhancer) %>% table() %>% as.data.frame() %>% dplyr::filter(Freq >1) %>% dplyr::rename(subjectHits='.')
  iTargetEnhancer <- iEnhType_enhancerBed[as.numeric(unfactor(i_summaryTab$subjectHits))]
  values(iTargetEnhancer) <- data.frame(siteCount = i_summaryTab$Freq)
  iTargetEnhancer.peakAnno <- ChIPseeker::annotatePeak(iTargetEnhancer, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame() 
  
  enhancer_buff <- rbind(data.frame(iTargetEnhancer.peakAnno, enhancerGroup=iEnhType), enhancer_buff)
  
  # ipathway <- ReactomePA::enrichPathway(iTargetEnhancer.peakAnno$geneId)
  # if(nrow(ipathway) >0) {
  #   print(paste0("Enhancers from which overlapped with multiple hyper methyled sites, are enriched in following pathways:", iEnhType))
  #   ipathway %>% as.data.frame() %>% dplyr::select(Description) %>% unlist() %>%  as.vector() %>% print()
  # }
}
# readr::write_tsv(enhancer_buff, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/multiple_HyperMethyled_in_enhancers.tsv"), progress = FALSE)
```


#### <h style="color:red"> enhancers have multiple hyper methyled  (LFC3) </h>

multiple HyperMethyled DMPs (**adj.P.Val < 0.01 & abs(logFC) >= 0.3**) in enhancers 

OUTPUT FILE: `multiple_HyperMethyled_in_enhancers_LFC3.tsv`

```{r message=FALSE, warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

DMP_data.sig %>% dplyr::filter(logFC >= 0.3)   %>% nrow()

HyperMethyled_LFC3.bed <- DMP_data.sig %>% dplyr::filter(logFC >= 0.3)  %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()

enhancer_buff_LFC3 <- data.frame()

for(iEnhType in c("7_EnhG1", "8_EnhG2", "9_EnhA1", "10_EnhA2", "11_EnhWk" )){
  iEnhType_enhancerBed <- enhancerBed[enhancerBed$name == iEnhType,]
  i_hyperDMP_in_enhancer <- GenomicRanges::findOverlaps(HyperMethyled_LFC3.bed, iEnhType_enhancerBed)
  i_summaryTab <- subjectHits(i_hyperDMP_in_enhancer) %>% table() %>% as.data.frame() %>% dplyr::filter(Freq >1) %>% dplyr::rename(subjectHits='.')
  iTargetEnhancer <- iEnhType_enhancerBed[as.numeric(unfactor(i_summaryTab$subjectHits))]
  values(iTargetEnhancer) <- data.frame(siteCount = i_summaryTab$Freq)
  iTargetEnhancer.peakAnno <- ChIPseeker::annotatePeak(iTargetEnhancer, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame() 
  
  enhancer_buff_LFC3 <- rbind(data.frame(iTargetEnhancer.peakAnno, enhancerGroup=iEnhType), enhancer_buff_LFC3)
}

# readr::write_tsv(enhancer_buff_LFC3, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/multiple_HyperMethyled_in_enhancers_LFC3.tsv"), progress = FALSE)
```

#### <h style="color:red"> enhancers have multiple hyper methyled  (LFC2) </h>

multiple HyperMethyled DMPs (**adj.P.Val < 0.01 & abs(logFC) >= 0.2**) in enhancers 

OUTPUT FILE: `multiple_HyperMethyled_in_enhancers_LFC2.tsv`

```{r message=FALSE, warning=FALSE}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

DMP_data.sig %>% dplyr::filter(logFC >= 0.2)   %>% nrow()

HyperMethyled_LFC2.bed <- DMP_data.sig %>% dplyr::filter(logFC >= 0.2)  %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()

enhancer_buff_LFC2 <- data.frame()

for(iEnhType in c("7_EnhG1", "8_EnhG2", "9_EnhA1", "10_EnhA2", "11_EnhWk" )){
  iEnhType_enhancerBed <- enhancerBed[enhancerBed$name == iEnhType,]
  i_hyperDMP_in_enhancer <- GenomicRanges::findOverlaps(HyperMethyled_LFC2.bed, iEnhType_enhancerBed)
  i_summaryTab <- subjectHits(i_hyperDMP_in_enhancer) %>% table() %>% as.data.frame() %>% dplyr::filter(Freq >1) %>% dplyr::rename(subjectHits='.')
  iTargetEnhancer <- iEnhType_enhancerBed[as.numeric(unfactor(i_summaryTab$subjectHits))]
  values(iTargetEnhancer) <- data.frame(siteCount = i_summaryTab$Freq)
  iTargetEnhancer.peakAnno <- ChIPseeker::annotatePeak(iTargetEnhancer, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame() 
  
  enhancer_buff_LFC2 <- rbind(data.frame(iTargetEnhancer.peakAnno, enhancerGroup=iEnhType), enhancer_buff_LFC2)
}

# readr::write_tsv(enhancer_buff_LFC2, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/multiple_HyperMethyled_in_enhancers_LFC2.tsv"), progress = FALSE)
```



#### annotate query enhancers to genes

OUTPUT FILE: `HyperMethyled_in_enhancers.tsv`

```{r}
enhancer_buff <- data.frame()

for(iEnhType in c("7_EnhG1", "8_EnhG2", "9_EnhA1", "10_EnhA2", "11_EnhWk" )){
  iEnhType_enhancerBed <- enhancerBed[enhancerBed$name == iEnhType,]
  i_hyperDMP_in_enhancer <- GenomicRanges::findOverlaps(HyperMethyled.bed, iEnhType_enhancerBed)
  i_summaryTab <- subjectHits(i_hyperDMP_in_enhancer)  %>% table() %>% as.data.frame()  %>% dplyr::rename(subjectHits='.')
  iTargetEnhancer <- iEnhType_enhancerBed[as.numeric(unfactor(i_summaryTab$subjectHits))]
  values(iTargetEnhancer) <- data.frame(siteCount = i_summaryTab$Freq)
  iTargetEnhancer.peakAnno <- ChIPseeker::annotatePeak(iTargetEnhancer, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame()
  
  enhancer_buff <- rbind(data.frame(iTargetEnhancer.peakAnno, enhancerGroup=iEnhType), enhancer_buff)
}
# readr::write_tsv(enhancer_buff, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/HyperMethyled_in_enhancers.tsv"), progress = FALSE)

```

#### hypo methyled sites on TssBiv, EnhBiv, ReprPC

OUTPUT FILE: `HypoMethyled_in_chromatinStates.tsv`

```{r}
HypoMethyled.bed <- DMP_data.sig.hypo %>% dplyr::mutate(CHR=paste0("chr", CHR))  %>% dplyr::mutate(start=MAPINFO) %>% dplyr::mutate(end=MAPINFO+1) %>% dplyr::select(CHR, start, end)  %>% GenomicRanges::GRanges()

hypo_buff <- data.frame()

for(istate in c("14_TssBiv", "15_EnhBiv", "16_ReprPC" )){
  istateBed <- ENCFF072SCA.bed[ENCFF072SCA.bed$name == istate,]
  i_hypoDMP_in_state <- GenomicRanges::findOverlaps(HypoMethyled.bed, istateBed)
  i_summaryTab <- subjectHits(i_hypoDMP_in_state) %>% table() %>% as.data.frame()  %>% dplyr::rename(subjectHits='.')

  iTargetRegions <- istateBed[as.numeric(unfactor(i_summaryTab$subjectHits))]
  values(iTargetRegions) <- data.frame(siteCount = i_summaryTab$Freq)
  iTargetRegions.peakAnno <- ChIPseeker::annotatePeak(iTargetRegions, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame()
  
  hypo_buff <- rbind(data.frame(iTargetRegions.peakAnno, state=istate), hypo_buff)
  
  # ipathway <- ReactomePA::enrichPathway(iTargetRegions.peakAnno$geneId)
  # if(nrow(ipathway) >0) {
  #   print(paste0("Regions which overlapped with hypo methyled sites, are enriched in following pathways:", istate))
  #   ipathway %>% as.data.frame() %>% dplyr::select(Description) %>% unlist() %>%  as.vector() %>% print()
  # }
}
# readr::write_tsv(hypo_buff, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/HypoMethyled_in_chromatinStates.tsv"), progress = FALSE)
```

## DMR in Monocyte chromatin states

Matilda used following to generate DMRs:

````
myDMR <- champ.DMR(beta=myNorm,
              pheno=annotation1$cluster_DNAmethyl_new,
              compare.group= c("mC_1","mC_2"),
              arraytype="EPIC",
              method = "Bumphunter",
              minProbes=7,    #only select DMRs contain more than minProbes (default = 7)
              adjPvalDmr=0.01, #(default = 0.05) all would take too much time and we decided for this cut off
             cores=2,
              ## following parameters are specifically for Bumphunter method.
              maxGap=300,     #maximum length for a DMR (default = 300)
              cutoff=NULL,
              pickCutoff=TRUE,
              smooth=TRUE,
              smoothFunction=loessByCluster,
              useWeights=FALSE,
              permutations=NULL,
              B=250,
              nullMethod="bootstrap")
````

OUTPUT FILE: `DMR_annotation.tsv`

```{r message=FALSE, warning=FALSE}
DMR_data <- readr::read_delim(paste0(CMML_dataPath, "Matilda_data/DMRs_mC1_mC2_20230904.csv"), delim=";",locale=locale(decimal_mark = ","), progress=FALSE )

nrow(DMR_data)

DMR_data.bed  <- DMR_data %>% dplyr::select(seqnames, start, end, strand)  %>% GenomicRanges::GRanges()

find_DMR_overlaps <- GenomicRanges::findOverlaps(DMR_data.bed, ENCFF072SCA.bed)

# DMR_data.bed[queryHits(find_DMR_overlaps)]
DMR_overlapped_states <- ENCFF072SCA.bed[(subjectHits(find_DMR_overlaps))]

DMR_overlapped_states.peakAnno <- ChIPseeker::annotatePeak(DMR_overlapped_states, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame()

table(DMR_overlapped_states.peakAnno$name)
# ReactomePA::enrichPathway(DMR_overlapped_states.peakAnno$geneId)

# DMR_data.peakAnno <- ChIPseeker::annotatePeak(DMR_data.bed, tssRegion=c(-3000, 3000),  TxDb=txdb, annoDb="org.Hs.eg.db", verbose =FALSE) %>% as.data.frame()
# ReactomePA::enrichPathway(DMR_data.peakAnno$geneId)

# readr::write_tsv(DMR_data.peakAnno, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/DMR_annotation.tsv"), progress = FALSE)
```

#### is any DMR overlap with enhancer states? 

is there any DMR overlapped with enhancers which have multiple DMP?

```{r}
multiple_HyperMethyled_in_enhancers.bed <- readr::read_tsv(paste0(CMML_dataPath, "resultTables_CMML_to_Matild/multiple_HyperMethyled_in_enhancers.tsv"), progress = FALSE) %>% GenomicRanges::GRanges()

DMR_data.bed[queryHits(GenomicRanges::findOverlaps(DMR_data.bed, multiple_HyperMethyled_in_enhancers.bed))]

multiple_HyperMethyled_in_enhancers.bed[subjectHits(GenomicRanges::findOverlaps(DMR_data.bed, multiple_HyperMethyled_in_enhancers.bed))] %>% as.data.frame() %>% dplyr::select(SYMBOL) %>% unlist() %>% as.vector() %>% unique()

```

### enhancer gene associations 

#### enhanceratlas

enhanceratlas is using `hg19`, need to left up to hg38 before overlapping with enhancers

```{r message=FALSE, warning=FALSE}
## read in data
enhanceratlas <- readr::read_tsv(paste0(CMML_dataPath, "enhanceratlas/enhanceratlas.org_data_AllEPs_hs_CD14+_EP.txt"), col_names = FALSE) %>%dplyr::rename(allInfo=X1, score=X2)

## add enhancer and gene info
enhanceratlas <- enhanceratlas  %>% dplyr::mutate( enhancer= unlist(lapply( .$allInfo, function(x) str_split( x, "_")[[1]][1] )), geneID= unlist( lapply( .$allInfo, function(x) stringr::str_split( stringr::str_split( x, "_")[[1]][2], "\\$")[[1]][1])), geneSymbol= unlist( lapply( .$allInfo, function(x) stringr::str_split( stringr::str_split( x, "_")[[1]][2], "\\$")[[1]][2]))) 
## add column for GRange build
enhanceratlas <- enhanceratlas %>% dplyr::mutate(chr=unlist(lapply(.$enhancer, function(x) stringr::str_split(x, ":")[[1]][1])), start=unlist(lapply(.$enhancer, function(x) stringr::str_split(stringr::str_split(x, ":")[[1]][2], "-")[[1]][1])), end=unlist(lapply(.$enhancer, function(x) stringr::str_split(stringr::str_split(x, ":")[[1]][2], "-")[[1]][2])))
## to GRanges object
enhanceratlas.GR <- enhanceratlas %>% GenomicRanges::GRanges()

enhancer.GR <- enhancer_buff %>%  GenomicRanges::GRanges()

# enhanceratlas.GR[subjectHits(GenomicRanges::findOverlaps(enhancer.GR, enhanceratlas.GR))] %>% as.data.frame() %>% dplyr::select(geneSymbol)  %>% unlist() %>% as.vector() %>% unique()
# 
# HyperMethyled_in_enhancers_gene <-  cbind(as.data.frame(enhancer.GR[queryHits(GenomicRanges::findOverlaps(enhancer.GR, enhanceratlas.GR))]), as.data.frame(enhanceratlas.GR[subjectHits(GenomicRanges::findOverlaps(enhancer.GR, enhanceratlas.GR))] ))
# 
# colnames(HyperMethyled_in_enhancers_gene)[c(1,2,3,20,21,22)] <- c("stateChr", "stateStart", "stateEnd", "enhancerChr", "enhancerStart", "enhancerEnd")
# 
# HyperMethyled_EnhInt <- HyperMethyled_in_enhancers_gene %>% dplyr::select(stateChr, stateStart, stateEnd, siteCount, enhancerGroup, enhancerChr, enhancerStart, enhancerEnd, enhancer, geneID, geneSymbol) %>% unique()
# 
# HyperMethyled_EnhInt %>% dplyr::select(geneSymbol)   %>% unlist() %>% as.vector() %>% unique()
```

#### GeneHancer

```{r}
geneHancer <- rtracklayer::import(paste0(CMML_dataPath, "GeneHancer/GeneHancer_v5.17.bed"))
geneHancer.gff <- readr::read_tsv(paste0(CMML_dataPath, "GeneHancer/GeneHancer_v5.17.gff"))

geneHancer.gff$enhancerid <- unlist(lapply(geneHancer.gff$attributes, function(x) gsub("genehancer_id=", "", stringr::str_split(x, ";")[[1]][1])))
```


```{r message=FALSE, warning=FALSE}
#### DMR set overlapping with enhancer states and have multiple DMP 
DMR_set01 <- DMR_data.bed[queryHits(GenomicRanges::findOverlaps(DMR_data.bed, multiple_HyperMethyled_in_enhancers.bed))] %>% unique()

DMR_enhancer.set01 <- cbind(as.data.frame(DMR_set01[queryHits(GenomicRanges::findOverlaps(DMR_set01, geneHancer))]), as.data.frame(geneHancer[subjectHits(GenomicRanges::findOverlaps(DMR_set01, geneHancer))])["name"]) %>% dplyr::rename(enhancerid=name) %>% merge(geneHancer.gff[c("enhancerid", "attributes")], by="enhancerid")

DMR_enhancer.set01$attributes <- gsub("genehancer_id=|connected_gene=", "", DMR_enhancer.set01$attributes)

DMR_enhancer.set01$attributes <- gsub(";score=", ":", DMR_enhancer.set01$attributes)

DMR_enhancer.set01
nrow(DMR_enhancer.set01)

#### all DMR
DMR_enhancer.set02 <- cbind(as.data.frame(DMR_data.bed[queryHits(GenomicRanges::findOverlaps(DMR_data.bed, geneHancer))]), as.data.frame(geneHancer[subjectHits(GenomicRanges::findOverlaps(DMR_data.bed, geneHancer))])["name"])  %>% dplyr::rename(enhancerid=name) %>% merge(geneHancer.gff[c("enhancerid", "attributes")], by="enhancerid")

DMR_enhancer.set02$attributes <- gsub("genehancer_id=|connected_gene=", "", DMR_enhancer.set02$attributes)
DMR_enhancer.set02$attributes <- gsub(";score=", ":", DMR_enhancer.set02$attributes)

head(DMR_enhancer.set02)
# readr::write_tsv(DMR_enhancer.set02, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/DMR_gene_associations.tsv"), progress = FALSE)

# DMR_data.bed[queryHits(GenomicRanges::findOverlaps(DMR_data.bed, geneHancer))]
# geneHancer[subjectHits(GenomicRanges::findOverlaps(DMR_data.bed, geneHancer))]
# 
# geneHancer[subjectHits(GenomicRanges::findOverlaps(enhancer.GR, geneHancer))]
```


```{r message=FALSE, warning=FALSE}
HyperMethyled_in_enhancers.bed <- readr::read_tsv(paste0(CMML_dataPath, "resultTables_CMML_to_Matild/HyperMethyled_in_enhancers.tsv"), progress = FALSE) %>% GenomicRanges::GRanges()

HyperMethyled_in_enhancers <- cbind(as.data.frame(HyperMethyled_in_enhancers.bed[queryHits(GenomicRanges::findOverlaps(HyperMethyled_in_enhancers.bed, geneHancer))]), as.data.frame(geneHancer[subjectHits(GenomicRanges::findOverlaps(HyperMethyled_in_enhancers.bed, geneHancer))])["name"]) %>% dplyr::rename(enhancerid=name) %>% merge(geneHancer.gff[c("enhancerid", "attributes")], by="enhancerid")

HyperMethyled_in_enhancers$attributes <- gsub("genehancer_id=|connected_gene=", "", HyperMethyled_in_enhancers$attributes)

HyperMethyled_in_enhancers$attributes <- gsub(";score=", ":", HyperMethyled_in_enhancers$attributes)

head(HyperMethyled_in_enhancers)

# readr::write_tsv(HyperMethyled_in_enhancers, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/HyperMethyled_in_enhancers_gene_associations.tsv"), progress = FALSE)
```

**LFC>=0.3**

```{r message=FALSE, warning=FALSE}
LFC3_HyperMethyled_in_enhancers.bed <- readr::read_tsv(paste0(CMML_dataPath, "resultTables_CMML_to_Matild/multiple_HyperMethyled_in_enhancers_LFC3.tsv"), progress = FALSE) %>% GenomicRanges::GRanges()

LFC3_HyperMethyled_in_enhancers <- cbind(as.data.frame(LFC3_HyperMethyled_in_enhancers.bed[queryHits(GenomicRanges::findOverlaps(LFC3_HyperMethyled_in_enhancers.bed, geneHancer))]), as.data.frame(geneHancer[subjectHits(GenomicRanges::findOverlaps(LFC3_HyperMethyled_in_enhancers.bed, geneHancer))])["name"]) %>% dplyr::rename(enhancerid=name) %>% merge(geneHancer.gff[c("enhancerid", "attributes")], by="enhancerid")

LFC3_HyperMethyled_in_enhancers$attributes <- gsub("genehancer_id=|connected_gene=", "", LFC3_HyperMethyled_in_enhancers$attributes)
LFC3_HyperMethyled_in_enhancers$attributes <- gsub(";score=", ":", LFC3_HyperMethyled_in_enhancers$attributes)

# readr::write_tsv(LFC3_HyperMethyled_in_enhancers, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/HyperMethyled_in_enhancers_gene_associations_LFC3.tsv"), progress = FALSE)
```


**LFC>=0.2**

```{r message=FALSE, warning=FALSE}
LFC2_HyperMethyled_in_enhancers.bed <- readr::read_tsv(paste0(CMML_dataPath, "resultTables_CMML_to_Matild/multiple_HyperMethyled_in_enhancers_LFC2.tsv"), progress = FALSE) %>% GenomicRanges::GRanges()

LFC2_HyperMethyled_in_enhancers <- cbind(as.data.frame(LFC2_HyperMethyled_in_enhancers.bed[queryHits(GenomicRanges::findOverlaps(LFC2_HyperMethyled_in_enhancers.bed, geneHancer))]), as.data.frame(geneHancer[subjectHits(GenomicRanges::findOverlaps(LFC2_HyperMethyled_in_enhancers.bed, geneHancer))])["name"]) %>% dplyr::rename(enhancerid=name) %>% merge(geneHancer.gff[c("enhancerid", "attributes")], by="enhancerid")

LFC2_HyperMethyled_in_enhancers$attributes <- gsub("genehancer_id=|connected_gene=", "", LFC2_HyperMethyled_in_enhancers$attributes)
LFC2_HyperMethyled_in_enhancers$attributes <- gsub(";score=", ":", LFC2_HyperMethyled_in_enhancers$attributes)

# readr::write_tsv(LFC2_HyperMethyled_in_enhancers, file=paste0(CMML_dataPath, "resultTables_CMML_to_Matild/HyperMethyled_in_enhancers_gene_associations_LFC2.tsv"), progress = FALSE)
```


```{r}
# ggbio::tracks(ggbio::autoplot(geneHancer[1]))
# ggbio::tracks(ggbio::autoplot(txdb[1]))
# data(genesymbol, package = "biovizBase")
# wh <- genesymbol[c("TMEM204")]
# wh <- range(wh, ignore.strand = TRUE)
# library(Homo.sapiens)
# class(Homo.sapiens)
# library(BSgenome.Hsapiens.UCSC.hg38)
# class(BSgenome.Hsapiens.UCSC.hg38)
# 
# ggbio::tracks(ggbio::autoplot(BSgenome.Hsapiens.UCSC.hg38, which=genesymbol[c("TMEM204")]))
# 
# library(BSgenome.Hsapiens.UCSC.hg19)
# data(genesymbol, package = "biovizBase")
# p1 <- ggbio::autoplot(Hsapiens, which = resize(genesymbol["ACTB"], width = 50))
# p2 <- ggbio::autoplot(Hsapiens, which = resize(genesymbol["ALDOA"], width = 50), geom = "rect")
# ggbio::tracks(text = p1, rect = p2)

# p.ideo <- ggbio::Ideogram(genome = "hg19")
# gr17 <- GRanges("chr17", IRanges(41234415, 41234569))
# library(VariantAnnotation)
# fl.vcf <- system.file("extdata", "17-1409-CEU-brca1.vcf.bgz", package="biovizBase")
# vcf <- readVcf(fl.vcf, "hg19")
# vr <- as(vcf[, 1:3], "VRanges")
# vr <- renameSeqlevels(vr, value = c("17" = "chr17"))
# p.vr <- ggbio::autoplot(vr, which = wh)
# gr.txdb <- biovizBase::crunch(txdb, which = wh)
# p.txdb <- ggbio::autoplot(Homo.sapiens, which = wh)
# tks <- ggbio::tracks(p.ideo,  dbSNP = p.vr, gene = p.txdb,heights = c( 3, 3,  4)) + ggbio::theme_tracks_sunset()
# tks
```


## play with metadata

```{r message=FALSE, warning=FALSE}
CMML_metadata <- readr::read_delim(paste0(CMML_dataPath, "Matilda_data/Meta_Data_20230831.csv"), progress = FALSE, delim = ";", quote = '"') 

CMML_metadata <- as.data.frame(t(CMML_metadata[!is.na(CMML_metadata[,c("TET2")]),]) )# filter out one patient without mutation data available

colnames(CMML_metadata) <- CMML_metadata["id",]

CML_mutation <- (CMML_metadata[c("TET2", "ASXL1", "SRSF2", "U2AF1", "PRPF8", "SF3B1", "ZRSR2"),])

ComplexHeatmap::Heatmap(type_convert(CML_mutation), name="mutation", show_column_names = FALSE, col=structure(1:2, names = c("0", "1")), cluster_columns = TRUE, column_split  = as.vector(unlist(CMML_metadata["cluster_WXS",])),  top_annotation = HeatmapAnnotation(cohort=as.vector(unlist(CMML_metadata["cohort",])), sex = as.vector(unlist(CMML_metadata["sex",])), WHO.class = as.vector(unlist(CMML_metadata["WHO.class",])), WXS = as.vector(unlist(CMML_metadata["cluster_WXS",])), FAB.group = as.vector(unlist(CMML_metadata["FAB.group",])), WBC = as.numeric(unlist(CMML_metadata["WBC",])), cluster_DNAmethyl = as.vector(unlist(CMML_metadata["cluster_DNAmethyl",])), TET2_multihit = as.vector(unlist(CMML_metadata["TET2_multihit",])), TET2_multihit_CNVadj = as.vector(unlist(CMML_metadata["TET2_multihit_CNVadj",])), del4q=  as.vector(unlist(CMML_metadata["del4q",])), TET2_SRSF2 = as.vector(unlist(CMML_metadata["TET2_SRSF2",])), show_legend = FALSE ) )

```


## session info

```{r message=FALSE, warning=FALSE, echo=FALSE, eval=FALSE}
sessionInfo()
```

